src/xml_parser_main_metadata_together.c

```diff
@@ -337,133 +351,215 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
                                         gboolean allow_out_of_order,
                                         GError **err)
 {
-    int ret = CRE_OK;
-    CR_FILE *primary_f = NULL;
-    CR_FILE *filelists_f = NULL;
-    CR_FILE *other_f = NULL;
-    GError *tmp_err = NULL;
+    assert(pkgcb || newpkgcb);
+    cr_PkgIterator* pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, newpkgcb, newpkgcb_data, warningcb, warningcb_data, allow_out_of_order, err
+    );
 
-    cr_CbData cbdata;
-    cbdata.in_progress_pkgs_list = NULL;
-    cbdata.in_progress_pkgs_hash = NULL;
-    cbdata.newpkgcb = newpkgcb;
-    cbdata.newpkgcb_data = newpkgcb_data;
-    cbdata.pkgcb = pkgcb;
-    cbdata.pkgcb_data = pkgcb_data;
+    // TODO: do this more properly
+    assert(pkg_iterator);
+    assert(!*err);
 
-    if (allow_out_of_order) {
-        cbdata.in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    cr_Package* package = NULL;
+    GError* tmp_err = NULL;
+
+    while (package = cr_PkgIterator_parse_next(pkg_iterator, err)) {
+        // call user package callback
+        if (pkgcb && pkgcb(package, pkgcb_data, &tmp_err)) {
+            if (tmp_err) {
+                g_propagate_prefixed_error(err, tmp_err, "Parsing interrupted: ");
+            } else {
+                g_set_error(err, ERR_DOMAIN, CRE_CBINTERRUPTED, "Parsing interrupted");
+            }
+            // cr_package_free(package);   // breaks test_xml_parser_main_metadata_together_pkgcb_abort
+            cr_PkgIterator_free(pkg_iterator, err);
+            return (*err)->code;
+        } else {
+            // If callback return CRE_OK but it simultaneously set
+            // the tmp_err then it's a programming error.
+            assert(tmp_err == NULL);
+        };
+    }
+    assert(cr_PkgIterator_is_finished(pkg_iterator) || !*err);
```

Comment:
Same here. I think we can easily run into the assert when there is some problem during parsing for example.
```suggestion
    if (*err) {
        return (*err)->code;
    }
    assert(cr_PkgIterator_is_finished(pkg_iterator));
```

Reply:
We're already checking this two lines down, do we need both?  I will go ahead and remove the `!*err` from the assert however

Reply:
Unless we add the error check there as well, some of the tests are failing for me (running into the `assert(cr_PkgIterator_is_finished(pkg_iterator))`):

`test_cr_xml_parse_main_metadata_together_01_out_of_order_pkgs`
`test_cr_xml_parse_main_metadata_together_04_newpkgcb_interrupt`
`test_cr_xml_parse_main_metadata_together_07_warningcb_interrupt`

So I think we want it there to properly report when there is an error/interupt.